### ISSUE TITLE
Epic: Move-Stable Row Ids

### ISSUE THREAD

**Issue by wjones127** [lancedb/lance#2307]
## Motivation

When we compaction data files, the row id changes. This causes us to need to update the index files whenever we compact. When the index files are updated, it invalidates them in the cache, degrading query performance. If row ids were stable when rows were moved, this would not happen.

## Scope

This epic makes row ids stable after **moving**. It does not make them stable after updates. Rows that are updated will be deleted and appended under new ids.

A future epic will cover "primary keys", which will be the point at which row ids will be stable after updates in addition to moves. This is kept out of scope for now to keep the workload of this manageable.

## Design

In very simple terms:

1. Add row ids as auto-incrementing u64 id. The manifest will track `max_row_id` and assign in similar process as fragment ids are assigned during the commit loop.
2. Each fragment metadata will contain a small row id index. This index maps from row id to row address. (Row address is what we currently call `_rowid`.) In most cases, such as after an append, this will be a simple range of values `(max_row_id + 1)..(physical_rows + max_row_id + 1)`.
3. ~Deletion files will be superceded by tombstones contained in the row id index. This cuts down on total number of files to manage.~
4. A new feature flag will be introduced to make sure older readers don't try to interpret these new row ids.

## Plan

* [x] #2308
* [x] Manifest changes for stable row id (feature flag + fields) #2363
  * Overwrite
  * Update
  * Create table
* [x] Scan with row id and address https://github.com/lancedb/lance/pull/2441
* [x] Take with stable row id https://github.com/lancedb/lance/pull/2447
* [x] Create index, query (including prefilter) https://github.com/lancedb/lance/pull/2452
* [x] Compaction and querying https://github.com/lancedb/lance/pull/2544

The following tasks have been moved into the [Primary Keys](https://github.com/lancedb/lance/issues/2454) epic:
* [ ] Follow ups for stabilization
   * [ ] Replace custom bitmap implementation
   * [ ] Finalize serialization format
   * [ ] Optimize row id access given real benchmarks
* [ ] External files and cleanup
  * Write out external files if large enough
  * Cleanup implementation

## Week of August 12

* [x] Benchmark prefilter performance with stable row id https://github.com/lancedb/lance/pull/2706
* [x] Move on to [#2454](https://github.com/lancedb/lance/issues/2454)

**Comment by wjones127** [comment]
## Week of July 29

Compaction and querying is waiting on review.

* [ ] Compaction and querying https://github.com/lancedb/lance/pull/2544

**Comment by wjones127** [comment]
Closing this, and continuing work in the #2454 epic.
